File: app/securesocial/provider/providers/GitHubProvider.java
/**
 * Copyright 2011 Ealden Esto E. Escanan (ealden at gmail dot com) - twitter: @ealden
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package securesocial.provider.providers;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.apache.commons.lang.StringUtils;
import play.Logger;
import play.Play;
import play.libs.WS;
import securesocial.provider.AuthenticationException;
import securesocial.provider.OAuth2Provider;
import securesocial.provider.ProviderType;
import securesocial.provider.SocialUser;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class GitHubProvider extends OAuth2Provider {
    private static final String GITHUB_ORGANIZATION = "securesocial.github.organization";

    private static final String AUTHENTICATED_USER = "https://api.github.com/user?access_token=%s";
    private static final String AUTHENTICATED_USER_ORGS = "https://api.github.com/user/orgs?access_token=%s";

    private static final String LOGIN = "login";
    private static final String NAME = "name";
    private static final String PICTURE = "avatar_url";
    private static final String EMAIL = "email";

    private static final String ERROR_MESSAGE = "message";

    public GitHubProvider() {
        super(ProviderType.github);
    }

    @Override
    protected void fillProfile(SocialUser user, Map<String, Object> authContext) {
        WS.HttpResponse response = WS.url(AUTHENTICATED_USER, user.accessToken).get();

        if (response.success()) {
            populateSocialUserInfo(user, response);
        } else {
            handleError(response);
        }
    }

    private void populateSocialUserInfo(SocialUser user, WS.HttpResponse response) {
        JsonObject authenticatedUser = response.getJson().getAsJsonObject();

        String organization = (String) Play.configuration.get(GITHUB_ORGANIZATION);

        if (StringUtils.isNotBlank(organization)) {
            validateUserMembership(user, organization);
        }

        user.id.id = authenticatedUser.get(LOGIN).getAsString();

        JsonElement displayName = authenticatedUser.get(NAME);

        if (displayName != null) {
            user.displayName = displayName.getAsString();
        } else {
            user.displayName = user.id.id;
        }

        JsonElement picture = authenticatedUser.get(PICTURE);

        if (picture != null) {
            user.avatarUrl = picture.getAsString();
        }

        JsonElement email = authenticatedUser.get(EMAIL);

        if (email != null) {
            user.email = email.getAsString();
        }
    }

    private void handleError(WS.HttpResponse response) {
        Integer status = response.getStatus();

        JsonObject error = response.getJson().getAsJsonObject();
        String message = error.get(ERROR_MESSAGE).getAsString();

        Logger.error("Error retrieving user information. Status: %i, message: %s.", status, message);

        throw new AuthenticationException();
    }

    private void validateUserMembership(SocialUser user, String organization) {
        WS.HttpResponse response = WS.url(AUTHENTICATED_USER_ORGS, user.accessToken).get();

        if (response.success()) {
            validateUserMembership(organization, response);
        } else {
            handleError(response);
        }
    }

    private void validateUserMembership(String organization, WS.HttpResponse response) {
        JsonArray userOrganizations = response.getJson().getAsJsonArray();
        List<String> organizations = extractOrganizations(userOrganizations);

        if (!organizations.contains(organization)) {
            Logger.error("User not part of, or user membership not public for, organization: %s.", organization);

            throw new AuthenticationException();
        }
    }

    private List<String> extractOrganizations(JsonArray userOrganizations) {
        List<String> organizations = new ArrayList<String>();

        Iterator<JsonElement> i = userOrganizations.iterator();

        while (i.hasNext()) {
            JsonObject userOrganization = i.next().getAsJsonObject();
            String organization = userOrganization.get(LOGIN).getAsString();

            organizations.add(organization);
        }

        return organizations;
    }
}

File: module-code/app/securesocial/core/providers/GitHubProvider.scala
/**
 * Copyright 2012-2014 Jorge Aliss (jaliss at gmail dot com) - twitter: @jaliss
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package securesocial.core.providers

import play.api.libs.ws.Response
import securesocial.core._
import securesocial.core.services.{CacheService, RoutesService}

import scala.concurrent.Future

/**
 * A GitHub provider
 *
 */
class GitHubProvider(routesService: RoutesService,
                     cacheService: CacheService,
                     client: OAuth2Client)
  extends OAuth2Provider(routesService, client, cacheService)
{
  val GetAuthenticatedUser = "https://api.github.com/user?access_token=%s"
  val AccessToken = "access_token"
  val TokenType = "token_type"
  val Message = "message"
  val Id = "id"
  val Name = "name"
  val AvatarUrl = "avatar_url"
  val Email = "email"

  override val id = GitHubProvider.GitHub

  override protected def buildInfo(response: Response): OAuth2Info = {
    val values: Map[String, String] = response.body.split("&").map( _.split("=") ).withFilter(_.size == 2)
        .map( r => (r(0), r(1)))(collection.breakOut)
    val accessToken = values.get(OAuth2Constants.AccessToken)
    if ( accessToken.isEmpty ) {
      logger.error(s"[securesocial] did not get accessToken from $id")
      throw new AuthenticationException()
    }
    OAuth2Info(
      accessToken.get,
      values.get(OAuth2Constants.TokenType),
      values.get(OAuth2Constants.ExpiresIn).map(_.toInt),
      values.get(OAuth2Constants.RefreshToken)
    )
  }

  def fillProfile(info: OAuth2Info): Future[BasicProfile] = {
    import scala.concurrent.ExecutionContext.Implicits.global
    client.retrieveProfile(GetAuthenticatedUser.format(info.accessToken)).map { me =>
        (me \ Message).asOpt[String] match {
          case Some(msg) =>
            logger.error(s"[securesocial] error retrieving profile information from GitHub. Message = $msg")
            throw new AuthenticationException()
          case _ =>
            val userId = (me \ Id).as[Int]
            val displayName = (me \ Name).asOpt[String]
            val avatarUrl = (me \ AvatarUrl).asOpt[String]
            val email = (me \ Email).asOpt[String].filter(!_.isEmpty)
            BasicProfile(id, userId.toString, None, None, displayName, email, avatarUrl, authMethod, oAuth2Info = Some(info))
        }
    } recover {
      case e: AuthenticationException => throw e
      case e  =>
        logger.error( "[securesocial] error retrieving profile information from github", e)
        throw new AuthenticationException()
    }
  }
}

object GitHubProvider {
  val GitHub = "github"
}
